### PR TITLE
Allow variable height in grid rows

### PR DESCRIPTION
**Pull request by cnrudd**
This is a WIP for this issue: #2196

Hoist P/R Checklist
-------------------

**Pull request authors:** Review and check off the below. Items that do not apply can also be
checked off to indicate they have been considered. If unclear if a step is relevant, please leave
unchecked and note in comments.

- [x] Caught up with `develop` branch as of last change.
- [x] Added CHANGELOG entry, or determined not required.
- [x] Reviewed for breaking changes, added `breaking-change` label + CHANGELOG if so.
- [x] Updated doc comments / prop-types, or determined not required.
- [ ] Reviewed and tested on Mobile, or determined not required.
- [x] Created Toolbox branch / PR, or determined not required.

**If your change is still a WIP**, please use the "Create draft pull request" option in the split
button below to indicate it is not ready yet for a final review.

> **Pull request reviewers:** when merging this P/R, please consider using a *squash commit* to
> collapse multiple intermediate commits into a single commit representing the overall feature
> change. This helps keep the commit log clean and easy to scan across releases. PRs containing a
> single commit should be *rebased* when possible.

### CODE REVIEW

**Comment by TomTirapani** [comment]
This branch works fine, but as discussed there are a couple of issues that unfortunately I haven't been able to work around.

### Sizing Modes.
I've scoured the agGrid docs, and there doesn't seem to be any way to get `Column.agOptions.autoHeight: true` to work with sizing modes.

If your grid has a `getRowHeight()` function **defined** then autoHeight is ignored, and our grids rely on the default `getRowHeight()` method to apply the sizing modes. Your grid becomes locked into 'standard' sizing mode if any column uses autoHeight.

I've looked to see if there is an API hook we could use to return the height from *within* the `getRowHeight()` function (and then `min()` with the sizing mode height), but no such hook seems to exists. The `AutoHeightCalculator` util that agGrid uses in its source is not exported.

The only approach I could think of would be to extend our autosize service to be able to determine height. It would render a placeholder component, and determine its height given a specified width. We could then use this within the `getRowHeight()` function to return the minimum height for the row. At that point we are effectively taking agGrid out of the equation. That is likely to be a considerable amount of work, so wanted to check if we think it is worth the investment.


### Autosize
Less of an issue than sizing modes, but column autosize will stretch out the column width so that its content is all in one line. While I think this behaviour is **correct**, it may be surprising and lead to _very_ wide columns if `autosizeMaxWidth` is not used. We could consider disabling autosize for columns with `Column.agOptions.wrapText: true`, similar to how we disable it for flex columns.


### Conclusion
At this point, I don't feel comfortable supporting auto row height as a "first-class" member of the grid API unless we can resolve the sizingMode conflict above. If this is required for a client app, this PR allows us to support it as a "hidden" option via `Column.agOptions` as long as the developer is willing to accept the caveats above.

**Comment by TomTirapani** [comment]
Okay -- I managed to get this working while respecting sizing modes.

There is a hook on the `agApi` we can use to calculate the row height, even when the getRowHeight function is implemented. This allows us to compare the auto height against the sizing mode height and choose the largest.

Unfortunately, I'm having difficult getting it to work with DataViews (or rather, with elementRenderers in general), will continue to investigate...

**Comment by cnrudd** [comment]
@TomTirapani,

Awesome that you found the solution to the autoheight on grids.
I've reached the same point/issue on dataview.  AgGrid is Ignoring the elementRender output when calculating autoheight.
Maybe a brief chat tomorrow?

**Comment by cnrudd** [comment]
It looks like the cause of the autoHeight fail for cellRenderers that are React Components is that the agGrid team has not yet developed a solution for psuedo rendering reactjs cell renderer components to get their height.

Here are two variations of the same ag-grid plunker demo, in which I modified the cellRenderer component for the currency column to repeat the currency 3 times in divs.  This should produce a tall cell that expands the rowHeight.
It works in Angular, but does not in React.
Angular: https://plnkr.co/plunk/TPGICc4PNVV2cAUm
ReactJS: https://plnkr.co/plunk/lVwzyPGgDvChQPo0

When I step through the code in the Chrome debugger when running the React version, I see in this file:
webpack:///./node_modules/@ag-grid-community/core/dist/es6/rendering/row/autoHeightCalculator.js?a87b
image
some code that tries to render an Angular component, but does not try to render a React component.

When filtering for autoHeight in the changelog (https://www.ag-grid.com/ag-grid-changelog/),
the entries that come up suggest they have been working through some autoHeight issues for a few versions now.
image

**Comment by cnrudd** [comment]
I found this bug discussed in this issue: https://github.com/ag-grid/ag-grid/issues/4138
TLDR:  it is being tracked in their [pipeline](https://www.ag-grid.com/ag-grid-pipeline/) as AG-4724.
It is not yet fixed, it is marked as "backlog".

**Comment by TomTirapani** [comment]
Thanks for the research Colin.

Disappointing to learn that this is an issue on AgGrid's side, and that its been ongoing for a few versions now with no fix in sight.

I think for the `Grid` and `AgGrid` use cases, this PR should help as long as we make it clear that it doesn't work for `elementRenderers` (which we discourage anyway). We can open this PR and merge if somebody else (@lbwexler?) wants to sanity check it.

For the `DataView` case, unfortunately it doesn't look like variable row heights are going to be an option with AgGrid. That leaves us with a few options:

+ Allow `DataViewModel` to accept a `renderer` in addition to an `elementRenderer`, and render html in my client app. We could make this a requirement of using auto height. Devs will have to decide between using an `elementRenderer` with a fixed item height (good for most cases), or a `renderer` for variable row height.

+ Allow `DataViewModel.itemHeight` to be a closure that receives the record. Developers *may* be able to use this some heuristic to return a different height based on the record's content. Not as good as actually sizing it, but may be sufficient for some cases.

+ If all else fails, I can change my client app to display a vanilla scrollable list of components rather than a `DataView`. I'd lose the buffered rendering and would have to manually implement the grouping, but in this particular use-case there shouldn't be too many rows for performance to suffer.

**Comment by cnrudd** [comment]
@TomTirapani Hi Tom,  I took a run on this idea:
"Allow DataViewModel.itemHeight to be a closure "
and got the dataview to work in these new branches:
https://github.com/xh/hoist-react/tree/dataViewAutoHeight
https://github.com/xh/toolbox/tree/dataViewAutoHeight

It's not totally clean.  Probably can be improved.  Happy to discuss tomorrow.

**Comment by cnrudd** [comment]
@TomTirapani Hi Tom,
Here are new branches that demo an approach that adds the dataviewItem to the DOM to get the height:
https://github.com/xh/toolbox/tree/dataViewReactDomRender
https://github.com/xh/hoist-react/tree/dataViewReactDomRender

This preserves the ag-Grid virttualization feature.

**Comment by cnrudd** [comment]
Todos from CR with Lee:

1. make autoHeight:true a HoistGridColumn prop that sets agOptions{autoHeight:true, wrapText:true}
2. add warning if a column has autoHeight:true and an elementRenderer component (won't work)
3. move comment test to a test grid in Admin
